test/test_optim.py

```diff
@@ -22,12 +22,19 @@
     optim_db, optims, OptimizerErrorEnum, _get_optim_inputs_including_global_cliquey_kwargs, TensorTracker)
 from torch.testing._internal.common_device_type import (
     instantiate_device_type_tests, largeTensorTest, onlyCPU, onlyCUDA, skipMPS, TEST_WITH_ROCM, onlyNativeDeviceTypes)
-from torch.testing._internal.common_utils import markDynamoStrictTest, parametrize, run_tests, TestCase
+from torch.testing._internal.common_utils import markDynamoStrictTest, parametrize, run_tests, TestCase, TEST_WITH_TORCHDYNAMO
 from torch.testing._internal.common_cuda import _create_scaling_case
 from torch.testing._internal.common_dtype import floating_types_and
 
 FP16_REDUCED_PRECISION = {'atol': 1e-5, 'rtol': 1e-4}
 
+def with_inductor_if_compiling(fn):
+    """ For peak memory tests, use inductor instead of eager backend """
+    if TEST_WITH_TORCHDYNAMO:
+        return torch.compile(fn, dynamic=False)
```

Comment:
Is the default backend here inductor?

What was the cause of the discrepancy with eager as backend anyway? Shouldn't eager compare just fine with eager in terms of memory usage? Or does dynamo use more on top of what eager does?

Reply:
Oh, if the test passes when we check capturable on the param group I'd be happier with that change which would assert that dynamo doesn't use extra memory. We could have another test testing what you describe about memory regressions in test_compiled_optimizer too.

Reply:
So I looked at this more deeply - it still fails even when I condition on capturable in the param group. However it looks like the main culprit here is that the dynamo single tensor eager backend uses *less* memory than eager, as a result, since the max memory comparison at the bottom uses the single tensor to compute the expected result the estimate is lower. Perhaps I add another intermediate in the case where we are compiling for the computation?

Reply:
lol do you know how much less memory dynamo uses? if it’s deterministic let’s just add that in (vs an entire intermediate)

Reply:
So one other issue - Adagrad actually does allocate more in foreach because of how we don't call item(), we basically allocate many small tensors in the scalar math sections. I will add an extra intermediate specifically in the cases where this happens.

Reply:
yeah it looks like its about 6k bytes, also though, there does appear to be an issue with NAdam and RAdam, where the eager backend doesn't seem to take into account the `del`s. Is it okay if a file an issue and increase the memory on those two optimizers when we are tracing with dynamo?

Reply:
Is it eager or dynamo that doesn't take into account the dels? Yes, feel free to file an issue and link it in the code as rationale.

Reply:
It's dynamo (torch FX)